vnext/JSI/Universal/ChakraRtRuntime.cpp

```diff
@@ -0,0 +1,129 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#ifndef CHAKRACORE
+
+#include "ChakraRtRuntime.h"
+
+#include "ByteArrayBuffer.h"
+#include "ChakraPreparedJavaScript.h"
+#include "Unicode.h"
+#include "Utilities.h"
+
+#include <jsrt.h>
+
+namespace Microsoft::JSI {
+
+facebook::jsi::Value ChakraRtRuntime::evaluatePreparedJavaScript(
+    const std::shared_ptr<const facebook::jsi::PreparedJavaScript> &preparedJs) {
+  assert(preparedJs);
+
+  std::shared_ptr<const ChakraPreparedJavaScript> chakraPreparedJs =
+      std::static_pointer_cast<const ChakraPreparedJavaScript>(preparedJs);
+
+  JsValueRef result = nullptr;
+
+  std::wstring sourceCodeUtf16 = Common::Unicode::Utf8ToUtf16(
+      Common::Utilities::CheckedReinterpretCast<const char *>(chakraPreparedJs->SourceCode()->data()),
+      chakraPreparedJs->SourceCode()->size());
+
+  std::wstring sourceUrlUtf16 = Common::Unicode::Utf8ToUtf16(chakraPreparedJs->SourceUrl());
+
+  // We have to use a const_cast here due to JsRunSerializedScript taking in a
+  // BYTE* instead of a const BYTE*.
+  JsErrorCode error = JsRunSerializedScript(
+      sourceCodeUtf16.c_str(),
+      static_cast<BYTE *>(const_cast<uint8_t *>(chakraPreparedJs->Bytecode()->data())),
+      JS_SOURCE_CONTEXT_NONE /*sourceContext*/,
+      sourceUrlUtf16.c_str(),
+      &result);
+
+  // Fall back to evaluating source code if bytecode evaluation fails.
+  if (error == JsErrorBadSerializedScript) {
+    JsValueRef exception = nullptr;
+    VerifyJsErrorElseThrow(JsGetAndClearException(&exception));
+    return EvaluateJavaScriptImpl(sourceCodeUtf16.c_str(), sourceUrlUtf16.c_str());
+  }
+
+  VerifyJsErrorElseThrow(error);
+  return ToJsiValue(ChakraObjectRef(result));
+}
+
+std::string ChakraRtRuntime::description() {
+  return "ChakraRtRuntime";
+}
+
+// TODO (yicyao): Chakra does not support weak reference semantics. We should
+// implement weak reference through std::shared_ptr<ChakraObjectRef>. For now,
+// we throw exceptions in createWeakObject and lockWeakObject to prevent
+// possible memory leaks.
+
+facebook::jsi::WeakObject ChakraRtRuntime::createWeakObject(const facebook::jsi::Object &object) {
+  throw facebook::jsi::JSINativeException("ChakraRtRuntime does not support createWeakObject().");
+}
+
+facebook::jsi::Value ChakraRtRuntime::lockWeakObject(const facebook::jsi::WeakObject &weakObject) {
+  throw facebook::jsi::JSINativeException("ChakraRtRuntime does not support lockWeakObject().");
+}
+
+facebook::jsi::JSRuntimeVersion_t ChakraRtRuntime::Version() {
+  // TODO (yicyao): Implement this function.
```

Comment:
>// TODO (yicyao): Implement this function. [](start = 2, length = 42)

I tried searching for a C++ UWP API for getting dll version, but I could not find any. I'll revisit this later.

Reply:
Couldwe make this noisy if it's called instead of returning garbage?

---
In reply to: [347681819](https://github.com/microsoft/react-native-windows/pull/3677#discussion_r347681819) [](ancestors = 347681819)